Source/AppIntegrationVerification.swift

```diff
@@ -0,0 +1,146 @@
+                case .echo(let token):
+                    return URL(string: Self.baseURLString.appending("echo?btn_ref=\(token)"))!
+                case .getToken(let token):
+                    return URL(string: Self.baseURLString.appending("get-token?btn_ref=\(token)"))!
+                case .testpostInstall(let success):
+                    return URL(string: Self.baseURLString.appending("test-post-install?success=\(success)"))!
+                case .version(let version):
+                    return URL(string: Self.baseURLString.appending("version?version=\(version)"))!
+                }
+            }
+        }
+        
+        init?(url: URL) {
+            guard let rawValue = Self.commandFrom(url: url) else {
+                return nil
+            }
+            self.init(rawValue: rawValue)
+        }
+    }
+    
+    var application: UIApplicationType
+    var buttonDefaults: ButtonDefaultsType
+    var token: String {
+        buttonDefaults.attributionToken ?? "null"
+    }
+    
+    required init(application: UIApplicationType, defaults: ButtonDefaultsType) {
+        self.application = application
+        self.buttonDefaults = defaults
+    }
+    
+    func handleIncomingURL(_ url: URL) {
+        guard let command = Command(url: url) else {
+            return
+        }
+        
+        switch command {
+        case .echo:
+            reportResult(url: Command.Response.echo(token: token).url)
+        case .getToken:
+            reportResult(url: Command.Response.getToken(token: token).url)
+        case .testPostInstall:
+            reportResult(url: Command.Response.testpostInstall(success: buttonDefaults.hasFetchedPostInstallURL).url)
+        case .version:
+            reportResult(url: Command.Response.version(version: Version.stringValue).url)
+        case .quit:
+            reportResult(url: Command.Response.quit.url) { success in
+                if success {
+                    self.application.quit()
+                }
+            }
+        }
+    }
+    
+    func reportResult(url: URL, completion: ((Bool) -> Void)? = nil) {
+        if #available(iOS 10.0, *) {
+            application.open(url, options: [:], completionHandler: completion)
+        }
+    }
+}
+
+extension AppIntegrationVerification.Command {
+    
+    static func commandFrom(url: URL) -> String? {
+        if isEchoCommand(url: url) {
+            return Self.echo.rawValue
+        }
+        
+        if url.host == Self.testHost,
+              url.path.starts(with: "/action/") {
+            return url.path.components(separatedBy: "/").last
+        }
+        
+        return nil
+    }
+    
+    static func isEchoCommand(url: URL) -> Bool {
+        guard let urlComponents = URLComponents(string: url.absoluteString),
+              let echoItem = urlComponents.queryItems?.first(where: { $0.name == Self.echo.rawValue }),
+              ((echoItem.value as NSString?)?.boolValue ?? false) else {
```

Comment:
So many ?????? 😆

Reply:
It _might_ be an `NSString`, that _might_ be a `Bool` that should otherwise be `false`